### PR TITLE
support for true batches with multipack

### LINKED ISSUE #1082: BatchSamplerDataCollatorForSeq2Seq does not properly construct batches
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Batches should be of shape `(micro_batch_size, sequence_len)`

### Current behaviour

Batches are of shape `(1, micro_batch_size*sequence_len)`

### Steps to reproduce

Add `print('Out features shape', features['input_ids'].shape)` in `DataCollatorForSeq2Seq`  before `return features`, add `print("Batch size", len(features)); print("Len of first batch", len(features[0]['input_ids']))` in `BatchSamplerDataCollatorForSeq2Seq` before `return super().__call__(features, return_tensors=return_tensors)`.

Run `accelerate launch -m axolotl.cli.train examples/llama-2/fft_optimized.yml --micro_batch_size 4 --val_set_size 0`

Results in

```
Batch size 1
Len of first batch 16258
Out features shape torch.Size([1, 16384])
```

### Config yaml

_No response_

### Possible solution

If this is indeed not intended behavior I am happy to fix it

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.12

### axolotl branch-commit

main/d69ba2b

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by winglian**
# Description

Fixes #1082 

Constructs true batches when using multipack without flash attention.

This supports multipack with `eager_attention: true`, though it only seems optimal with `micro_batch_size: 1`.

## Motivation and Context

Flash attention needs a batch size of 1 when using multipack, but this PR opens up the possibility of using 4d attention masks with multipack and sdpa.

## How has this been tested?

added unit tests

### CODE REVIEW

**Comment by winglian** [comment]
@RicardoDominguez

**Comment by winglian** [comment]
~~it's unclear why this is 3x slower when using micro_batch_size > 1.~~ also, the scaled dot product attention still results in 0.0 loss.

EDIT: https://github.com/OpenAccess-AI-Collective/axolotl/pull/1230/commits/fce6ea70ab870c7a4b64b896370dd2e86ef180a8 fixes the speed issue

**Comment by winglian** [comment]
actually, the loss is stable when using fp16 with sdp_attention (for llama)
